### ISSUE TITLE
Desktop widget cropping photos in a bad way

### ISSUE THREAD

**Issue by mvodanovic** [deckerst/aves#1285]
**Describe the bug**
The widget to show photos crops photos vertically. I haven't found a way to show entire photos, they are always cropped. This worked fine for me until maybe a month ago, then it just broke. Maybe it's related to a major Android update? In the attached screenshot you can see the widget area (the widget is selected) and then the photo inside.

**To Reproduce**
1. Add the photo frame widget and have it display a photo.
2. The picture is cropped vertically. No changing of the widget size helps.

**Expected behavior**
The whole photo should be shown in the widget area, scaled and not cropped.

**Screenshots**
![Screenshot_20241110-190245](https://github.com/user-attachments/assets/07766ecd-42d2-4c68-9fd9-3574c9d2f5c5)

**System information and logs:**
Package: deckers.thibault.aves
Installer: com.android.vending
Aves version: 1.11.17-play, build 136
Flutter: stable 3.24.4
Android version: 15, API 35
Android build: AP3A.241105.007
Device: Google Pixel 7
Support: dynamic colors=true, geocoder=true, HDR=true
Mobile services: ready
Connectivity: wifi
System locales: en_US, en_SE
Storage volumes: /storage/emulated/0/
Storage grants: /storage/emulated/0/Pictures/, /storage/emulated/0/Android/media/
Error reporting: true

[aves-logs.txt](https://github.com/user-attachments/files/17692510/aves-logs.txt)

**Additional context**
None.

**Comment by deckerst** [comment]
> Maybe it's related to a major Android update?

Maybe related to switching from Android 14 to Android 15, but the home/launcher app itself is also an important piece of the puzzle, because it's the launcher app that informs Aves of the available size for the widget. Anyway, thanks for the report, I'll take a look.

**Comment by deckerst** [comment]
I can't reproduce it on a Pixel 7 emulator with Android 15, and the attached logs are nearly empty, so I'm not sure how to approach this.

Do you use a custom launcher? How do you get a grid with 7 apps in a row? (with the emulator the max is 5).

Maybe, the widget takes time to resize because something goes wrong or gets delayed during the widget initialization. How many items are in your collection, more or less?

**Comment by mvodanovic** [comment]
Hey! Yes, I'm using the Nova launcher. It might be related to that. If Nova is not reporting something correctly to you, then I guess you can't do much about it. 😕 Thanks for checking anyway!

There are not that many images in the collection that I filter for. I basically display only images that contain a certain tag. And I've tagged maybe 20 images max. I guess that shouldn't be the issue?

**Comment by deckerst** [comment]
> I guess that shouldn't be the issue?

Probably not. I'll give it a try with Nova Launcher and see whether there's a workaround.

**Comment by deckerst** [comment]
I just tried Nova Launcher with the default settings on my device with Android 14 and it works as expected: 

https://github.com/user-attachments/assets/7840659c-e61a-45ce-9a80-7124f2393798

Could you identify some specific Nova Launcher settings that make it fail?

**Comment by e-t-l** [comment]
@deckerst Could we consider reopening this issue? I am encountering the exact same bug:

Pixel 8 Pro
Android 15 (GrapheneOS)
Nova Launcher 8.1
Aves Libre 1.11.19

The widget used to work fine when I had Aves Libre (undoubtesly an older version), Nova 8.1, and Android 14 on my old Pixel 7, which I replaced last year. I only recently tried making a photo widget on my current phone and realized it's buggy.

I'm happy to collect logs or perform troubleshooting on my device to help fix this bug.

**Comment by deckerst** [comment]
@e-t-l it's a combination of Android 15 and Nova Launcher. It seems Nova doesn't give the available size the same way as standard home launchers.

Giving details about your Nova settings (the app grid size, etc) and the system info from `About` > `Bug Report` could help.

**Comment by e-t-l** [comment]
Oh heck that was easy. @deckerst the bug is a regression introduced in Aves **version 1.11.5**

I went back and tested out older versions (using the Izzy apks, so it wouldn't conflict with the Libre apk I currently have installed as my main gallery app). The bug does not exist in v1.11.4 but it does in v1.11.5+onwards.

Were any changes made to the code that handles the photo frame widget in that version?

**Comment by deckerst** [comment]
Yes, the code changed, but it was to support Android 15. With a standard launcher the new code works fine. So that's why I say it's also Nova specific.

**Comment by e-t-l** [comment]
Oh I didn't see that you already replied to my first comment! My grid size is 7x4. For the bug report, are you asking for something generated by Nova or by the Android system?

**Comment by deckerst** [comment]
From Aves > About > Bug Report > System info.

**Comment by e-t-l** [comment]
[aves-logs-20250120_231858.txt](https://github.com/user-attachments/files/18486168/aves-logs-20250120_231858.txt)

Here you go! That's from the fresh install of v1.11.5, where Aves has all default settings.

**Comment by deckerst** [comment]
I could reproduce the issue with Nova launcher on my device with Android 15. It was fine before with Nova when I was still on Android 14. I'll take a closer look when I find the time.

**Comment by deckerst** [comment]
After digging, I'm quite confident it's a bug with Nova Launcher.

In modern Android versions, the widget host (i.e. the launcher app) can provide different sizes, and the app can provide a layout for each of these sizes. In practice the host gives different sizes matching different situations: landscape/portrait (and 2 more for foldables). So the app provides the different layouts once, then the host selects among the layouts when the user rotates their screen (if home screen rotation is enabled).

On the app side, I'm following this:
https://developer.android.com/develop/ui/views/appwidgets/layouts#provide-exact-layouts

On the host side, it should select the layout that best fits the available size. Standard launcher does that well, whatever the screen orientation, whatever the widget sizes. But Nova always picks up the layout with the smallest area, not the best fit.

For reference on the layout selection:
https://developer.android.com/reference/android/widget/RemoteViews#RemoteViews(java.util.Map%3Candroid.util.SizeF,android.widget.RemoteViews%3E)

**Comment by e-t-l** [comment]
That's fair. I'll submit a bug report with Nova. In the meantime, would it be possible to put a hotfix in Aves, like a settings toggle labeled, "Fix photoframe dimensions," with some very simple logic like:
```
if (hotfixtoggle) 
    {use legacy photoframe code from v1.11.4} 
else 
    {use latest photoframe code}
```

**Comment by deckerst** [comment]
I won't include such a toggle for the sake of a buggy app. But if you find an open-source launcher app that handles resizable widgets, and Aves also has the issue with that app, I'm interested. I've tried various launchers but nearly none of them support widgets, or they do not support resizing them.

**Comment by Nanoseb** [comment]
I wanted to mention that this problem also occurs with Lawnchair. But also that the app "material photo widget" does work properly. Maybe worth having a look at its source to see how they do it: https://github.com/fibelatti/photo-widget
